### PR TITLE
Translate empty row rule and reduction to apply! interface

### PR DESCRIPTION
**Pull request by joehuchette**

### CODE REVIEW

File: src/lp/empty_row.jl [hunk 1]
```diff
@@ -1,21 +1,57 @@
...
+## Infeasibility checks
+The problem is infeasible if row $i$ is empty and either 1) its lower bound is positive ($l_{i} > ϵ$) or 2) its upper bound is negative ($u_{i} < -ϵ$).
+
+# Postsolve
+
+## Misc
+
+* This is a primal reduction.
+* This reduction is non-destructive.
+* THis reduction does not create any fill-in.
```
Comment (joehuchette) [inline src/lp/empty_row.jl:24]: *this

File: src/lp/empty_row.jl [hunk 2]
```diff
@@ -1,21 +1,57 @@
...
+for prescribed tolerance $ϵ$.
+
+## Presolve
+
+If row $i$ is empty, the corresponding constraint is completely removed from the problem.
+
+## Infeasibility checks
+The problem is infeasible if row $i$ is empty and either 1) its lower bound is positive ($l_{i} > ϵ$) or 2) its upper bound is negative ($u_{i} < -ϵ$).
+
+# Postsolve
```
Comment (joehuchette) [inline src/lp/empty_row.jl:18]: ## Postsolve (and fill in)

File: src/presolve.jl [hunk 3]
```diff
@@ -647,19 +628,6 @@ function round_integer_bounds!(ps::PresolveData{T}) where {T}
     return nothing
 end
 
-"""
-    remove_fixed_variables!(ps::PresolveData)
-
-Remove all fixed variables.
-"""
-function remove_fixed_variables!(ps::PresolveData{T}) where {T}
```
Comment (joehuchette) [inline src/presolve.jl:655]: I believe this is no longer necessary after #7

File: src/lp/empty_row.jl [hunk 4]
```diff
@@ -1,21 +1,71 @@
...
+@doc raw"""
+    RemoveEmptyRow <: AbstractRule
 
+Eliminate a single row if it is empty, i.e. all variable coefficients are zero.
+In particular, consider row $i$ given as
+```math
+l^{r}_{i} ≤ \sum_{j} a_{i, j} x_{j} ≤ u^{r}_{i}.
+```
+We eliminate row $i$ if, for each variable index $j$,
+``math
```
Comment (mtanneau) [inline src/lp/empty_row.jl:10]:
```suggestion
```math
```

File: src/lp/empty_row.jl [hunk 5]
```diff
@@ -1,21 +1,71 @@
...
+lower bound is positive ($l^{r}_{i} > ϵ$) or 2) its upper bound is
+negative ($u^{r}_{i} < -ϵ$). In these two cases, a Farkas infeasibility
+certificate is given by setting $y$ to all zeros except for either
+$y\_lower_{i} = 1$ or $y\_upper_{i} = 1$, respectively.
+
+## Postsolve
+
+Sets the dual variable $y\_lower_{i} = \max\{0,y\}$ for the greater-than
+constraint and the dual variable $y\_upper_{i} = \max\{0,-y\}$ for the
+less-than constraint.
```
Comment (mtanneau) [inline src/lp/empty_row.jl:32]: Small clarification.
Unless row `i` is infeasible, both dual multipliers `y_l` and `y_u` are zero: this corresponds to the reduction `EmptyRow(i, zero(T))`, which is created in the non-infeasible case.

Thus, the `y_l = max(0, y)`, `y_u = max(0, -y)` in the post-solve only yield non-zero values if the row was infeasible.
My rationale for initially doing so was to have "status-agnostic" reductions and postsolve.

File: src/lp/empty_row.jl [hunk 6]
```diff
@@ -1,21 +1,71 @@
...
+Eliminate a single row if it is empty, i.e. all variable coefficients are zero.
+In particular, consider row $i$ given as
+```math
+l^{r}_{i} ≤ \sum_{j} a_{i, j} x_{j} ≤ u^{r}_{i}.
+```
+We eliminate row $i$ if, for each variable index $j$,
+``math
+| a_{i,j} | ≤ ϵ
+```
+for prescribed tolerance $ϵ$.
```
Comment (mtanneau) [inline src/lp/empty_row.jl:13]: I don't see this being done in the code, is it on purpose?

Comment (joehuchette) [reply src/lp/empty_row.jl]: Hmm no, I wrote this first and then never followed through on it. Do we want to do this? If so, at what point? It feels a bit weird to do it in ``load_problem!``: if the user is passing us the matrix with small nonzeros, it seems like we should respect that.

File: src/lp/empty_row.jl (hunk 5 above)
Comment (joehuchette) [reply src/lp/empty_row.jl]: Updated the doc, let me know what you think.

File: src/lp/empty_row.jl (hunk 6 above)
Comment (mtanneau) [reply src/lp/empty_row.jl]: If we do it, I would rather put it in a different rule, with a drop tolerance for rounding small values to zero and large ones to infinity.
As for the where, once at the beginning of presolve seems reasonable. See also https://github.com/mtanneau/MathOptPresolve.jl/issues/12#issuecomment-751526797

File: src/lp/empty_row.jl [hunk 7]
```diff
@@ -1,21 +1,73 @@
...
+If row $i$ is empty, the corresponding constraint is completely removed from
+the problem.
+
+## Infeasibility checks
+
+The problem is infeasible if row $i$ is empty and either 1) its
+lower bound is positive ($l^{r}_{i} > ϵ$) or 2) its upper bound is
+negative ($u^{r}_{i} < -ϵ$). In these two cases, a Farkas infeasibility
+certificate is given by setting $y$ to all zeros except for either
+$y\_lower_{i} = 1$ or $y\_upper_{i} = 1$, respectively.
```
Comment (mtanneau) [inline src/lp/empty_row.jl:22]: Documenter renders this in a single 3-line paragraph, which I find hard to read at first glance.
I'd suggest something like

```
For a prescribed tolerance $ϵ ≥ 0$, the problem is infeasible if either
1. row $i$ is empty and its lower bound is positive, i.e., $l^{r}_{i} > ϵ$, or
2. row $i$ is empty and its upper bound is negative, i.e., $u^{r}_{i} < -ϵ$.

The corresponding Farkas infeasibilities certificates are
1.  $y^{l} = e_{i}$, $y^{u} = 0$, $z^{l} = z^{u} = 0$,
2.  $y^{u} = e_{i}$, $y^{l} = 0$, $z^{l} = z^{u} = 0$,

where $e_{i}$ is the $i$-th vector of the canonical basis.
```
(it needs two spaces between the bullet point and the `$` to render correctly)

File: src/lp/empty_row.jl [hunk 8]
```diff
@@ -1,21 +1,73 @@
...
+
+## Postsolve
+
+If the problem is feasible, the corresponding dual variables  all dual
+multipliers $y\_lower_{i} = $y\_upper_{i} = 0$. If the problem is infeasible,
+we update the dual variables $y\_lower_{i} = \max\{0,y\}$ for the greater-than
+ constraint and the dual variable $y\_upper_{i} = \max\{0,-y\}$ for the
+ less-than constraint.
```
Comment (mtanneau) [inline src/lp/empty_row.jl:30]: We can shorten it to something like
```
If the problem is feasible, $y^{l}_{i} = y^{u}_{i} = 0$.

If the problem is infeasible, see the Farkas rays above.
```

It's also possible to move the Farkas certificates to this part of the docs. As you prefer.

**Review by mtanneau — CHANGES_REQUESTED** [review]
Can you add an entry in the docs as well?

**Review by mtanneau — APPROVED** [review]
LGTM, thanks!

**Pull request merged**